src/ai/backend/manager/scheduler/dispatcher.py

```diff
@@ -619,24 +628,95 @@ async def _schedule_single_node_session(
             # If sess_ctx.agent_id is already set for manual assignment by superadmin,
             # skip assign_agent_for_session().
             agent = sess_ctx.main_kernel.agent_row
-            agent_id: AgentId
+            agent_id: AgentId | None = None
             if agent is not None:
                 agent_id = agent.id
             else:
-                # Let the scheduler check the resource availability and decide the target agent
-                cand_agent = scheduler.assign_agent_for_session(
-                    compatible_candidate_agents, sess_ctx
-                )
-                if cand_agent is None:
-                    raise InstanceNotAvailable(
-                        extra_msg=(
-                            "Could not find a contiguous resource region in any agent "
-                            "big enough to host the session "
-                            f"(id: {sess_ctx.id}, resource group: {sess_ctx.scaling_group_name})"
-                        ),
+                sorted_agents = sorted(compatible_candidate_agents, key=lambda agent: agent.id)
+
+                if scheduler.sgroup_opts.roundrobin:
+                    rr_state_str: str | None = await sched_ctx.registry.shared_config.get_raw(
+                        "roundrobin_state"
                     )
-                assert cand_agent is not None
-                agent_id = cand_agent
+
+                    if rr_state_str is not None:
+                        rr_state: Dict[str, Any] = json.loads(rr_state_str)
+
+                        if requested_architecture in rr_state:
+                            schedulable_group_id = get_hash(
+                                "#".join(list(map(lambda agent: agent.id, sorted_agents)))
+                            )
+
+                            if schedulable_group_id == cast(
+                                Dict[str, Any], rr_state.get(requested_architecture)
+                            ).get("schedulable_group_id"):
+                                rr_index = cast(
+                                    int,
+                                    cast(Dict[str, Any], rr_state.get(requested_architecture)).get(
+                                        "next_index"
+                                    ),
+                                )
```

Comment:
This is just my opinion.
It is not bad to handle the branches through chaining methods. It looks like functional programming.
But I am not sure it is good in our code base cause I have never seen such `cast` chaining and I am worried about readability.
I just personally like to use walrus operator in such case.
```python
if (rr_state_data := rr_state.get(requested_architecture)) is not None:
    sg_id = rr_state_data.get("schedulable_group_id")
```
This is just my opinion, so please do not feel pressure to change this code. Let's talk about it later!